File: frontend/src/api/services.ts
import { api } from './http'

export type ServiceStatus = {
  name: string
  active: boolean
  enabled: boolean
  active_state: string
}

export type WireguardSummary = {
  interface: string
  listen_port: string | null
  address: string | null
  peer_count: number
}

export type WireguardPeer = {
  public_key: string
  endpoint: string
  allowed_ips: string
  latest_handshake: number
  transfer_rx: number
  transfer_tx: number
  persistent_keepalive: number
}

export type PiholeStats = {
  total_queries: number
  blocked_queries: number
  updated_at: number
}

export const getWireguardStatus = () =>
  api.get<ServiceStatus>('/api/v1/services/wireguard/status')
export const startWireguard = () =>
  api.post<ServiceStatus>('/api/v1/services/wireguard/start')
export const stopWireguard = () =>
  api.post<ServiceStatus>('/api/v1/services/wireguard/stop')
export const restartWireguard = () =>
  api.post<ServiceStatus>('/api/v1/services/wireguard/restart')
export const getWireguardSummary = () =>
  api.get<WireguardSummary>('/api/v1/services/wireguard/summary')
export const getWireguardPeers = () =>
  api.get<WireguardPeer[]>('/api/v1/services/wireguard/peers')

export const getPiholeStatus = () =>
  api.get<ServiceStatus>('/api/v1/services/pihole/status')
export const startPihole = () =>
  api.post<ServiceStatus>('/api/v1/services/pihole/start')
export const stopPihole = () =>
  api.post<ServiceStatus>('/api/v1/services/pihole/stop')
export const restartPihole = () =>
  api.post<ServiceStatus>('/api/v1/services/pihole/restart')
export const getPiholeStats = () =>
  api.get<PiholeStats>('/api/v1/services/pihole/stats')

File: backend/app/api/routes/services.py
"""Service control routes for WireGuard and Pi-hole."""

from fastapi import APIRouter, Depends

from app.api.deps import require_admin
from app.core.config import get_settings
from app.schemas.services import (
    PiholeStatsResponse,
    ServiceStatusResponse,
    WireguardPeerResponse,
    WireguardSummaryResponse,
)
from app.services.pihole import PiholeService
from app.services.wireguard import WireguardService

router = APIRouter(prefix="/services", tags=["services"])


def _to_status_response(status) -> ServiceStatusResponse:
    return ServiceStatusResponse(
        name=status.name,
        active=status.active,
        enabled=status.enabled,
        active_state=status.active_state,
    )


@router.get("/wireguard/status", response_model=ServiceStatusResponse)
async def wireguard_status(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = WireguardService(get_settings())
    return _to_status_response(service.status())


@router.post("/wireguard/start", response_model=ServiceStatusResponse)
async def wireguard_start(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = WireguardService(get_settings())
    return _to_status_response(service.start())


@router.post("/wireguard/stop", response_model=ServiceStatusResponse)
async def wireguard_stop(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = WireguardService(get_settings())
    return _to_status_response(service.stop())


@router.post("/wireguard/restart", response_model=ServiceStatusResponse)
async def wireguard_restart(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = WireguardService(get_settings())
    return _to_status_response(service.restart())


@router.get("/wireguard/summary", response_model=WireguardSummaryResponse)
async def wireguard_summary(
    _: object = Depends(require_admin),
) -> WireguardSummaryResponse:
    service = WireguardService(get_settings())
    summary = service.config_summary()
    return WireguardSummaryResponse(**summary.__dict__)


@router.get("/wireguard/peers", response_model=list[WireguardPeerResponse])
async def wireguard_peers(
    _: object = Depends(require_admin),
) -> list[WireguardPeerResponse]:
    service = WireguardService(get_settings())
    peers = service.peers()
    return [WireguardPeerResponse(**peer.__dict__) for peer in peers]


@router.get("/pihole/status", response_model=ServiceStatusResponse)
async def pihole_status(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = PiholeService(get_settings())
    return _to_status_response(service.status())


@router.post("/pihole/start", response_model=ServiceStatusResponse)
async def pihole_start(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = PiholeService(get_settings())
    return _to_status_response(service.start())


@router.post("/pihole/stop", response_model=ServiceStatusResponse)
async def pihole_stop(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = PiholeService(get_settings())
    return _to_status_response(service.stop())


@router.post("/pihole/restart", response_model=ServiceStatusResponse)
async def pihole_restart(_: object = Depends(require_admin)) -> ServiceStatusResponse:
    service = PiholeService(get_settings())
    return _to_status_response(service.restart())


@router.get("/pihole/stats", response_model=PiholeStatsResponse)
async def pihole_stats(_: object = Depends(require_admin)) -> PiholeStatsResponse:
    service = PiholeService(get_settings())
    stats = service.stats()
    return PiholeStatsResponse(**stats.__dict__)
